File: leetcode/algorithms/541_reverse_string_2/main.js
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var reverseStr = function (s, k) {
  const sArray = s.split("");
  if (sArray.length < k) {
    return sArray.reverse().join("");
  }

  let result = "";

  for (let i = 0; i < s.length; i += 2 * k) {
    const stringK = sArray.slice(i, i + k).reverse().join("");
    const left = sArray.slice(i + k, i + 2 * k).join("");
    result += stringK + left;
  }

  return result;
};


// Best Solution
var bestSolution = function (s, k) {
  if (k > s.length) {
    return s.split("").reverse().join("");
  }

  const split = s.split("");

  // reverse the segment and put it back
  for (let i = 0; i < s.length; i += 2 * k) {
    const reverse = split.splice(i, k).reverse();
    split.splice(i, 0, ...reverse);
  }

  return split.join("");
};

File: leetcode/algorithms/541_reverse_string_2/main.dart
class ReverseStringII {
  void swap(List<String> sArray, int l, int r) {
    while (l < r) {
      String temp = sArray[l];
      sArray[l++] = sArray[r];
      sArray[r--] = temp;
    }
  }

  String reverseStr(String s, int k) {
    int n = s.length;
    int i = 0;
    List<String> sArray = s.split("");

    while (i < n) {
      int j = i + k - 1 < n - 1 ? i + k - 1 : n - 1;
      swap(sArray, i, j);
      i += 2 * k;
    }

    return sArray.join();
  }

  // Solution
  String solution(String s, int k) {
    List<String> result = s.split('');

    for (int i = 0; i < s.length; i += 2 * k) {
      int start = i;
      int end = (i + k - 1 < s.length) ? i + k - 1 : s.length - 1;
      while (start <= end) {
        String temp = result[start];
        result[start] = result[end];
        result[end] = temp;
        start++;
        end--;
      }
    }

    return result.join();
  }
}
